crates/chain/src/indexed_tx_graph.rs

```diff
@@ -0,0 +1,389 @@
+    pub fn apply_update(&mut self, update: TxGraph<A>) -> IndexedAdditions<A, I::Additions> {
+        let graph_additions = self.graph.apply_update(update);
+
+        let mut index_additions = I::Additions::default();
+        for added_tx in &graph_additions.tx {
+            index_additions.append(self.index.index_tx(added_tx));
+        }
+        for (&added_outpoint, added_txout) in &graph_additions.txout {
+            index_additions.append(self.index.index_txout(added_outpoint, added_txout));
+        }
+
+        IndexedAdditions {
+            graph_additions,
+            index_additions,
+        }
+    }
+
+    /// Insert a floating `txout` of given `outpoint`.
+    pub fn insert_txout(
+        &mut self,
+        outpoint: OutPoint,
+        txout: &TxOut,
+    ) -> IndexedAdditions<A, I::Additions> {
+        let mut update = TxGraph::<A>::default();
+        let _ = update.insert_txout(outpoint, txout.clone());
+        self.apply_update(update)
+    }
+
+    /// Insert and index a transaction into the graph.
+    ///
+    /// `anchors` can be provided to anchor the transaction to various blocks. `seen_at` is a
+    /// unix timestamp of when the transaction is last seen.
+    pub fn insert_tx(
+        &mut self,
+        tx: &Transaction,
+        anchors: impl IntoIterator<Item = A>,
+        seen_at: Option<u64>,
+    ) -> IndexedAdditions<A, I::Additions> {
+        let txid = tx.txid();
+
+        let mut update = TxGraph::<A>::default();
+        if self.graph.get_tx(txid).is_none() {
+            let _ = update.insert_tx(tx.clone());
+        }
+        for anchor in anchors.into_iter() {
+            let _ = update.insert_anchor(txid, anchor);
+        }
+        if let Some(seen_at) = seen_at {
+            let _ = update.insert_seen_at(txid, seen_at);
+        }
+
+        self.apply_update(update)
+    }
+
+    /// Insert relevant transactions from the given `txs` iterator.
+    ///
+    /// Relevancy is determined by the [`Indexer::is_tx_relevant`] implementation of `I`. Irrelevant
+    /// transactions in `txs` will be ignored.
+    ///
+    /// `anchors` can be provided to anchor the transactions to blocks. `seen_at` is a unix
+    /// timestamp of when the transactions are last seen.
+    pub fn insert_relevant_txs<'t, T: Iterator<Item = &'t Transaction>>(
+        &mut self,
+        txs: T,
+        anchors: impl IntoIterator<Item = A> + Clone,
+        seen_at: Option<u64>,
+    ) -> IndexedAdditions<A, I::Additions> {
+        txs.filter_map(|tx| match self.index.is_tx_relevant(tx) {
+            true => Some(self.insert_tx(tx, anchors.clone(), seen_at)),
+            false => None,
+        })
+        .fold(Default::default(), |mut acc, other| {
+            acc.append(other);
+            acc
+        })
+    }
+}
+
+impl<A: Anchor, I: OwnedIndexer> IndexedTxGraph<A, I> {
+    pub fn try_list_owned_txs<'a, C: ChainOracle + 'a>(
```

Comment:
I guess "owned" is not the best term to use. Any transaction that alters balances of spks that we own.

Reply:
Yeah this is stupid, the caller should just insert relevant transactions.

Reply:
Removed.

Reply:
Since relevant will be approximately the same as relevant this definition of ownership for most indexes I think the caller can just filter `is_relevant` to narrow down to these transactions.